experimental/segmenter/src/math_helper.rs

```diff
@@ -245,87 +239,74 @@ impl<'a> MatrixBorrowedMut<'a, 1> {
-        // for i in 0..b.dim().0 {
-        //     self.submatrix_mut::<1>(i).add_dot_2d(a, b.submatrix(i));
-        // }
-        let lhs = a.as_slice();
         for i in 0..n {
-            if let (Some(dest), Some(rhs)) = (
-                self.as_mut_slice().get_mut(i),
-                b.as_slice().get_subslice(i * m..(i + 1) * m),
-            ) {
-                *dest += unrolled_dot_1(lhs, rhs);
+            if let (Some(dest), Some(b_sub)) = (self.as_mut_slice().get_mut(i), b.submatrix::<1>(i))
+            {
+                *dest += unrolled_dot_2(a.as_slice(), b_sub.data);
             } else {
                 debug_assert!(false, "unreachable: dims checked above");
             }
         }
     }
 
-    /// Calculate the dot product of a and b, adding the result to self.
-    ///
-    /// Self should be _MxN_; `a`, _O_; and `b`, _MxNxO_.
-    pub fn add_dot_3d_2(&mut self, a: MatrixZero<1>, b: MatrixZero<3>) {
-        let m = a.dim();
-        let n = self.as_borrowed().dim().0 * self.as_borrowed().dim().1;
-        debug_assert_eq!(
-            m,
-            b.dim().2,
-            "dims: {:?}/{:?}/{:?}",
-            self.as_borrowed().dim(),
-            a.dim(),
-            b.dim()
-        );
-        debug_assert_eq!(
-            n,
-            b.dim().0 * b.dim().1,
-            "dims: {:?}/{:?}/{:?}",
-            self.as_borrowed().dim(),
-            a.dim(),
-            b.dim()
-        );
-        // Note: The following two loops are equivalent, but the second has more opportunity for
-        // vectorization since it allows the vectorization to span submatrices.
-        // for i in 0..b.dim().0 {
-        //     self.submatrix_mut::<1>(i).add_dot_2d(a, b.submatrix(i));
-        // }
-        let lhs = a.as_slice();
-        for i in 0..n {
-            if let (Some(dest), Some(rhs)) = (
-                self.as_mut_slice().get_mut(i),
-                b.as_slice().get_subslice(i * m..(i + 1) * m),
-            ) {
-                *dest += unrolled_dot_2(lhs, rhs);
-            } else {
-                debug_assert!(false, "unreachable: dims checked above");
-            }
+    pub fn sigmoid(&mut self, r: Range<usize>) -> Option<()> {
+        let slice = self.data.get_mut(r)?;
+        for i in &mut *slice {
+            *i = sigmoid(*i);
+        }
+        Some(())
+    }
+
+    pub fn tanh(&mut self, r: Range<usize>) -> Option<()> {
+        let slice = self.data.get_mut(r)?;
+        for i in &mut *slice {
+            *i = tanh(*i);
+        }
+        Some(())
+    }
+
+    pub fn convolve(
+        &mut self,
+        i: MatrixBorrowed<'_, 1>,
+        c: MatrixBorrowed<'_, 1>,
+        f: MatrixBorrowed<'_, 1>,
+    ) -> Option<()> {
+        for idx in 0..self.data.len() {
+            *self.data.get_mut(idx)? =
+                i.data.get(idx)? * c.data.get(idx)? + self.data.get(idx)? * f.data.get(idx)?
         }
```

Comment:
After compiling the crate with `RUSTFLAGS='-C opt-level=2 -Cllvm-args=--pass-remarks=.*vector.* -Cllvm-args=--pass-remarks-analysis=.*vector.*'` and extracting the remarks related to "math_helper.rs" 

<details>
<summary> llvm opt remarks for math_helper.rs at CL</summary>
<br> 

```
remark: experimental/segmenter/src/math_helper.rs:520:9: loop not vectorized: cannot prove it is safe to reorder floating-point operations
remark: experimental/segmenter/src/math_helper.rs:557:9: loop not vectorized: cannot prove it is safe to reorder floating-point operations
remark: experimental/segmenter/src/math_helper.rs:339:36: loop not vectorized: value that could not be identified as reduction is used outside the loop
remark: experimental/segmenter/src/math_helper.rs:339:36: loop not vectorized: could not determine number of loop iterations
remark: experimental/segmenter/src/math_helper.rs:346:65: loop not vectorized: instruction cannot be vectorized
remark: experimental/segmenter/src/math_helper.rs:339:17: the cost-model indicates that interleaving is not beneficial
remark: experimental/segmenter/src/math_helper.rs:339:17: vectorized loop (vectorization width: 4, interleaved count: 1)
remark: experimental/segmenter/src/math_helper.rs:346:40: loop not vectorized: could not determine number of loop iterations
```
math_helper.rs:520 -> inner loop of unrolled_dot_1 (not vectorized)
math_helper.rs:557 -> inner loop of unrolled_dot_2 (not vectorized)
math_helper.rs:339 -> convolve loop (vectorized) as the loop has no data dependencies across indices
math_helper.rs:346 -> mul_tanh loop (not vectorized) tanh is an unvectorizable function

</details>

I think the speedup over c4152c5d73db19730295b7089323d0af6035fa8a comes from a semi-vectorized version of https://github.com/unicode-org/icu4x/blob/main/experimental/segmenter/src/lstm.rs#L175-L191.

After removing the loops and only computing the add_dot product at both HEAD and CL, we see near similar performance.

<details>
<summary>Benchmark HEAD/CL only computing add_dot </summary>
<br>

```
Line Break/UTF8/Th/lstm time:   [246.39 µs 246.60 µs 246.85 µs]
                        change: [+3.1495% +3.3542% +3.5639%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 18 outliers among 100 measurements (18.00%)
  3 (3.00%) high mild
  15 (15.00%) high severe

Line Break/UTF16/Th/lstm
                        time:   [248.21 µs 248.50 µs 248.82 µs]
                        change: [+3.6773% +3.9000% +4.1699%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 18 outliers among 100 measurements (18.00%)
  1 (1.00%) low severe
  3 (3.00%) high mild
  14 (14.00%) high severe
```
</details>


Additionally if we use $approxtanh(x) = \frac{2}{1+e^-2x} -1 $, mul_tanh vectorizes. 
<details>
<summary> Benchmark with $approxtanh(x)$ </summary>
<br>

```
Line Break/UTF8/Th/lstm time:   [313.47 µs 313.75 µs 314.08 µs]
                        change: [-34.471% -34.320% -34.183%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 10 outliers among 100 measurements (10.00%)
  4 (4.00%) high mild
  6 (6.00%) high severe

Line Break/UTF16/Th/lstm
                        time:   [313.34 µs 313.47 µs 313.61 µs]
                        change: [-34.703% -34.552% -34.410%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 15 outliers among 100 measurements (15.00%)
  5 (5.00%) high mild
  10 (10.00%) high severe
```
</details>

This passes the segmenter test suite with `cargo test --all-features`, but I'm not certain if we should use this approximation as the absolute error $|tanh(x) - approxtanh(x)|$ is as big as $1e^-3$ for the test suite.

Reply:
OK, so as far as this PR is concerned, we believe that the perf comes primarily from the vectorization of the `convolve` function?

Hmm, I wonder if the _most_ efficient would be a memory layout like the one we currently have but where we carry the matrix in chunks of 4 so that we can vectorize those operations together; we get the benefits of both memory locality (which I found was faster) and vectorization (which you found was faster)